### PR TITLE
[SPARK-26936][SQL] Fix bug of insert overwrite local dir can not create temporary path in local staging directory

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?
Th environment of my cluster as follows:
```
OS:Linux version 2.6.32-220.7.1.el6.x86_64 (mockbuild@c6b18n3.bsys.dev.centos.org) (gcc version 4.4.6 20110731 (Red Hat 4.4.6-3) (GCC) ) #1 SMP Wed Mar 7 00:52:02 GMT 2012
Hadoop: 2.7.2
Spark: 2.3.0 or 3.0.0(master branch)
Hive: 1.2.1
```

My spark run on deploy mode yarn-client. 

If I execute the SQL `insert overwrite local directory '/home/test/call_center/' select * from call_center`, a HiveException will appear as follows:
`Caused by: org.apache.hadoop.hive.ql.metadata.HiveException: java.io.IOException: Mkdirs failed to create file:/home/xitong/hive/stagingdir_hive_2019-02-19_17-31-00_678_1816816774691551856-1/-ext-10000/_temporary/0/_temporary/attempt_20190219173233_0002_m_000000_3 (exists=false, cwd=file:/data10/yarn/nm-local-dir/usercache/xitong/appcache/application_1543893582405_6126857/container_e124_1543893582405_6126857_01_000011)
at org.apache.hadoop.hive.ql.io.HiveFileFormatUtils.getHiveRecordWriter(HiveFileFormatUtils.java:249)`
Current spark sql generate a local temporary path in local staging directory.The schema of local temporary path start with `file`, so the HiveException appears.
This PR change the local temporary path to HDFS temporary path, and use DistributedFileSystem instance copy the data from HDFS temporary path to local directory.
If Spark run on local deploy mode, 'insert overwrite local directory' works fine.
## How was this patch tested?

UT cannot support yarn-client mode.The test is in my product environment.

### CODE REVIEW

**Comment by maropu** [comment]
Can you add tests before runing tests in Jenkins?

**Comment by beliefer** [comment]
> Can you add tests before runing tests in Jenkins?
Existing unit tests

**Comment by maropu** [comment]
ok to test

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala [hunk 1]
```diff
@@ -86,20 +86,21 @@ case class InsertIntoHiveDirCommand(
...
-        qualifiedPath
+        val dfs = qualifiedPath.getFileSystem(hadoopConf)
+        (qualifiedPath, dfs)
       }
+    if (!fs.exists(writeToPath)) {
+      fs.mkdirs(writeToPath)
+    }
 
-    val tmpPath = getExternalTmpPath(sparkSession, hadoopConf, writeToPath)
+    // The temporary path must be a HDFS path but a local path.
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: Yes,this is a hdfs path.

**Comment by beliefer** [comment]
@maropu Please review this pr again.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 2]
```diff
@@ -581,6 +581,38 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
     }
   }
 
+  test("insert overwrite to not exists dir from hive metastore table") {
+    withTempDir { dir =>
+      val path = dir.toURI.getPath
+      val notExistsPath = s"${path}/src/result"
+
+      sql(s"INSERT OVERWRITE LOCAL DIRECTORY '${notExistsPath}' SELECT * FROM src where key < 10")
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: nit: `${notExistsPath}` -> `$notExistsPath `

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 3]
```diff
@@ -581,6 +581,38 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
...
+
+      sql(
+        s"""
+           |INSERT OVERWRITE LOCAL DIRECTORY '${notExistsPath}'
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: ditto

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 4]
```diff
@@ -581,6 +581,38 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
...
+         """.stripMargin)
+
+      // use orc data source to check the data of path is right.
+      withTempView("orc_source") {
+        sql(
+          s"""
+             |CREATE TEMPORARY VIEW orc_source
+             |USING org.apache.spark.sql.hive.orc
+             |OPTIONS (
+             |  PATH '${notExistsPath}'
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: ditto

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 5]
```diff
@@ -581,6 +581,38 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
...
+             |)
+           """.stripMargin)
+
+        checkAnswer(
+          sql("select * from orc_source"),
+          sql("select * from src where key < 10"))
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: You should set literals in the expected answer.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 6]
```diff
@@ -581,6 +581,38 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
     }
   }
 
+  test("insert overwrite to not exists dir from hive metastore table") {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: What's the difference from the `"insert overwrite to not exist local dir"` test?

**Comment by maropu** [comment]
cc: @dongjoon-hyun

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala [hunk 7]
```diff
@@ -86,20 +86,21 @@ case class InsertIntoHiveDirCommand(
     val jobConf = new JobConf(hadoopConf)
 
     val targetPath = new Path(storage.locationUri.get)
-    val writeToPath =
+    val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)
+    val (writeToPath: Path, fs: FileSystem) =
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: OK, I have adjust it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala [hunk 8]
```diff
@@ -86,20 +86,21 @@ case class InsertIntoHiveDirCommand(
...
+        val dfs = qualifiedPath.getFileSystem(hadoopConf)
+        (qualifiedPath, dfs)
       }
+    if (!fs.exists(writeToPath)) {
+      fs.mkdirs(writeToPath)
+    }
 
-    val tmpPath = getExternalTmpPath(sparkSession, hadoopConf, writeToPath)
+    // The temporary path must be a HDFS path, not a local path.
+    val tmpPath = getExternalTmpPath(sparkSession, hadoopConf, qualifiedPath)
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: If aim path is local, we still need to use a non-local path.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 2 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: OK, I have adjust it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 3 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: OK, I have adjust it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 4 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: OK, I have adjust it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 5 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: These code refers to the writing of other UT.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 6 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: One is used for test result correct or not, another is used for test the not exist path created or not.

Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: Can you run these tests in a single case?

Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: Yes,I can remove "insert overwrite to not exist local dir" test and retain another.

**Comment by beliefer** [comment]
There parent path `/tmp/noexistdir` exists or not?
What is the deploy mode on which your query running?

**Comment by maropu** [comment]
My test case was that `/tmp/noexistdir` didn't exist first.

**Comment by beliefer** [comment]
OK,I have supplement the deploy mode for the PR description.

**Comment by beliefer** [comment]
I have run these SQL you provided on local[*] deploy mode too,but another inconsistent behavior appear.
```
ls /tmp/noexistdir
ls: cannot access /tmp/noexistdir: No such file or directory

scala> sql("""create table t(c0 int, c1 int)""")
res0: org.apache.spark.sql.DataFrame = []
scala> spark.table("t").explain
== Physical Plan ==
HiveTableScan [c0#5, c1#6], HiveTableRelation `default`.`t`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [c0#5, c1#6]

scala> sql("""insert into t values(1, 1)""")
scala> sql("""select * from t""").show
+---+---+                                                                       
| c0| c1|
+---+---+
|  1|  1|
+---+---+

scala> sql("""insert overwrite local directory '/tmp/noexistdir/t' select * from t""")
res1: org.apache.spark.sql.DataFrame = [] 

ls /tmp/noexistdir/t/
/tmp/noexistdir/t

vi /tmp/noexistdir/t
  1 
```
'/tmp/noexistdir/t' is not a directory but a file.
Normally, temp path have result file and '_SUCCESS' file.
I doubt result file cover the '_SUCCESS' file by code as follows:
` fs.listStatus(tmpPath).foreach {
        tmpFile => fs.rename(tmpFile.getPath, writeToPath)
      }`

**Comment by beliefer** [comment]
Then I compile the master branch and deploy on my hadoop cluster.I get the inconsistent behavior again.
`ls /tmp/noexistdir
ls: cannot access /tmp/noexistdir: No such file or directory`

`Java HotSpot(TM) 64-Bit Server VM warning: Using the ParNew young collector with the Serial old collector is deprecated and will likely be removed in a future release
Spark context Web UI available at http://10.198.66.204:55326
Spark context available as 'sc' (master = local[*], app id = local-1551259036573).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.0-SNAPSHOT
      /_/
         
Using Scala version 2.12.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_131)
Type in expressions to have them evaluated.
Type :help for more information.

scala> sql("""select * from t""").show
+---+---+                                                                       
| c0| c1|
+---+---+
|  1|  1|
+---+---+


scala> sql("""insert overwrite local directory '/tmp/noexistdir/t' select * from t""")
res1: org.apache.spark.sql.DataFrame = []                                       

scala> `

`ll /tmp/noexistdir/t
-rw-r--r-- 1 xitong xitong 0 Feb 27 17:19 /tmp/noexistdir/t
vi /tmp/noexistdir/t
  1`

**Comment by beliefer** [comment]
The origin code of `InsertIntoHiveDirCommand.scala` not create parent path on local `FileSystem`.

**Comment by beliefer** [comment]
I have update the PR and associated JIRA.

**Comment by beliefer** [comment]
@maropu Could you review the PR again?Thanks.

**Comment by beliefer** [comment]
I have check the source of Hadoop `LocalFileSystem `.`LocalFileSystem` don't implement the method `rename`. `LocalFileSystem` extends `ChecksumFileSystem` and the latter implement the method `rename`.
The method `rename` of `ChecksumFileSystem` as follows:
```
  public boolean rename(Path src, Path dst) throws IOException {
    if (fs.isDirectory(src)) {
      return fs.rename(src, dst);
    } else {
      if (fs.isDirectory(dst)) {
        dst = new Path(dst, src.getName());
      }

      boolean value = fs.rename(src, dst);
      if (!value)
        return false;

      Path srcCheckFile = getChecksumFile(src);
      Path dstCheckFile = getChecksumFile(dst);
      if (fs.exists(srcCheckFile)) { //try to rename checksum
        value = fs.rename(srcCheckFile, dstCheckFile);
      } else if (fs.exists(dstCheckFile)) {
        // no src checksum, so remove dst checksum
        value = fs.delete(dstCheckFile, true); 
      }

      return value;
    }
  }
```
If target path is a directory, `ChecksumFileSystem` will move source file into target path.
If target path is not a directory, `ChecksumFileSystem` will rename source file to target file.
There exists a variable named `fs` that is a `RawLocalFileSystem`. `RawLocalFileSystem` will call the method `rename` of UNIXFileSystem or `WinNTFileSystem`.

**Comment by maropu** [comment]
Does this issue only happen in yarn-client mode?

**Comment by beliefer** [comment]
> Does this issue only happen in yarn-client mode?
This SQL will occur `org.apache.hadoop.hive.ql.metadata.HiveException: java.io.IOException: Mkdirs failed to create file:/` on yarn-client mode first.
This SQL occurs inconsistent behavior on local mode.

**Comment by maropu** [comment]
You need to first check the test you added in this pr... it still passes without your fix.

**Comment by beliefer** [comment]
> You need to first check the test you added in this pr... it still passes without your fix.

Maybe UT run on local and depend on operating system or other difference of environment.

**Comment by beliefer** [comment]
I get the OS of UT is Linux.

**Comment by beliefer** [comment]
@maropu I have tried to find out why UT and my spark behave differently when executing `insert overwrite local directory`. But I'm failed! According to the source code of `InsertIntoHiveDirCommand`, there no chance to create the target path that doesn't exist yet.Could you help me, find out the reason.

**Comment by beliefer** [comment]
@maropu I move the conversation of create no exists local path with `insert overwrite local dir` to PR `https://github.com/apache/spark/pull/23950`. The current PR only used for `Mkdirs failed to create file` in yarn-client mode.

**Comment by beliefer** [comment]
@maropu  Could you review this PR again, thanks!

**Comment by maropu** [comment]
Why did you remove the tests?

**Comment by beliefer** [comment]
> Why did you remove the tests?
Sorry, I can't find the reason why the feature 'insert overwrite local directory' has an inconsistent behavior in different environment. I have moved the conversation of create no exists local path with `insert overwrite local dir` to PR https://github.com/apache/spark/pull/23950. The current PR only used for `Mkdirs failed to create file` in yarn-client mode.

**Comment by beliefer** [comment]
@maropu Sorry, I can't find the reason why the feature 'insert overwrite local directory' has an inconsistent behavior in different environment. I have moved the conversation of create no exists local path with insert overwrite local dir to PR #23950. The current PR only used for Mkdirs failed to create file in yarn-client mode.

**Comment by beliefer** [comment]
cc @maropu @gatorsmile @dongjoon-hyun @janewangfb @cloud-fan
Could you review this PR, thanks!

**Comment by beliefer** [comment]
@srowen Could you review this PR, thanks!

**Comment by srowen** [comment]
It looks plausible, but I'm concerned that the problem can't be reproduced anywhere here.
You seem to be using a path that's meant to be local to be a path on distributed storage as the 'fix'; that doesn't sound right but maybe I'm misunderstanding

**Comment by beliefer** [comment]
> It looks plausible, but I'm concerned that the problem can't be reproduced anywhere here.
> You seem to be using a path that's meant to be local to be a path on distributed storage as the 'fix'; that doesn't sound right but maybe I'm misunderstanding

This bug occurs in `yarn-client` mode. I use a  path on distributed storage to replace a local path. because `HDFS` can not recognize the path start with `file:/`.

**Comment by srowen** [comment]
This doesn't sound like the right fix, as it would change behavior much more broadly than that. Why is a local file used in the first place here? is it user error or something farther up that's specifying the wrong dir?

**Comment by beliefer** [comment]
> This doesn't sound like the right fix, as it would change behavior much more broadly than that. Why is a local file used in the first place here? is it user error or something farther up that's specifying the wrong dir?

Using the local path is the original behavior of spark. When Spark execute the SQL `insert overwrite local directory` in `yarn-client` mode, original Spark will use `LocalFileSystem` to create temp path start with `file:/`. `DistributedFileSystem`  will write data to the temp path but recognize the path.

**Comment by beliefer** [comment]
@srowen 
The original behavior of spark SQL is:
generate data in distributed cluster -------> write data to local temp path ------>  move data in local temp path to local target path.
But data can not write to local temp path.
This PR is:
generate data in distributed cluster -------> write data to distributed temp path in ------> move data in 
distributed temp path to local target path.

**Comment by srowen** [comment]
You're changing all code paths though, ones that work. I am not clear that's valid. Why not work out why this happens only in yarn-client mode?

**Comment by beliefer** [comment]
> You're changing all code paths though, ones that work. I am not clear that's valid. Why not work out why this happens only in yarn-client mode?

I paste the full stack as follows:
`Driver stacktrace:
        at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1599)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1587)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1586)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1586)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
        at scala.Option.foreach(Option.scala:257)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:831)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1820)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1769)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1758)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:642)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2037)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:194)
        ... 36 more
Caused by: org.apache.spark.SparkException: Task failed while writing rows.
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:285)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:197)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:196)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
        at org.apache.spark.scheduler.Task.run(Task.scala:109)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.hadoop.hive.ql.metadata.HiveException: java.io.IOException: Mkdirs failed to create file:/home/xitong/hive/stagingdir_hive_2019-02-19_17-31-00_678_1816816774691551856-1/-ext-10000/_temporary/0/_temporary/attempt_20190219173233_0002_m_000000_3 (exists=false, cwd=file:/data10/yarn/nm-local-dir/usercache/xitong/appcache/application_1543893582405_6126857/container_e124_1543893582405_6126857_01_000011)
        at org.apache.hadoop.hive.ql.io.HiveFileFormatUtils.getHiveRecordWriter(HiveFileFormatUtils.java:249)
        at org.apache.spark.sql.hive.execution.HiveOutputWriter.<init>(HiveFileFormat.scala:123)
        at org.apache.spark.sql.hive.execution.HiveFileFormat$$anon$1.newInstance(HiveFileFormat.scala:103)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.newOutputWriter(FileFormatWriter.scala:367)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.execute(FileFormatWriter.scala:378)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:269)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:267)
        at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1411)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:272)
        ... 8 more
Caused by: java.io.IOException: Mkdirs failed to create file:/home/xitong/hive/stagingdir_hive_2019-02-19_17-31-00_678_1816816774691551856-1/-ext-10000/_temporary/0/_temporary/attempt_20190219173233_0002_m_000000_3 (exists=false, cwd=file:/data10/yarn/nm-local-dir/usercache/xitong/appcache/application_1543893582405_6126857/container_e124_1543893582405_6126857_01_000011)
        at org.apache.hadoop.fs.ChecksumFileSystem.create(ChecksumFileSystem.java:449)
        at org.apache.hadoop.fs.ChecksumFileSystem.create(ChecksumFileSystem.java:435)
        at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:928)
        at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:821)
        at org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat.getHiveRecordWriter(HiveIgnoreKeyTextOutputFormat.java:80)
        at org.apache.hadoop.hive.ql.io.HiveFileFormatUtils.getRecordWriter(HiveFileFormatUtils.java:261)
        at org.apache.hadoop.hive.ql.io.HiveFileFormatUtils.getHiveRecordWriter(HiveFileFormatUtils.java:246)
        ... 16 more`

**Comment by beliefer** [comment]
According to the error:
`FileFormatWriter` will create `SingleDirectoryWriteTask` and call the method `execute`.
The `SingleDirectoryWriteTask` runs on one of nodes in yarn cluster.
The original code use a temp path start with `file:/ `, but this path is not a valid path in this node. the temp path only valid in driver node.

**Comment by cloud-fan** [comment]
Will we hit this bug when we deploy spark in cluster? Seems to me it's not specific to yarn.

**Comment by beliefer** [comment]
> Will we hit this bug when we deploy spark in cluster? Seems to me it's not specific to yarn.

Yes, If we deploy spark in cluster, this bug will occurs.

**Comment by srowen** [comment]
That makes more sense if this isn't YARN-specific, but isn't this still using a local path as if it's remote, or am I misreading?

**Comment by beliefer** [comment]
> That makes more sense if this isn't YARN-specific, but isn't this still using a local path as if it's remote, or am I misreading?

The SQL start with `insert overwrite local directory` decides the target path is local.
There exists an intermediate process write result data to a temp path first. After the intermediate process, move result data from temp path to local target path.
The original code uses a local path as temp path.
This PR uses a distributed path as temp path.

**Comment by srowen** [comment]
OK it's probably that I don't know this code well. Maybe my question is this: when `isLocal == true`, `storage.locationUri` is a local path, so `targetPath` points to a local directory. It is now used as a path on distributed storage here. It may not exist or be usable that way. Is that OK? It may happen to work but seems a little wrong. Or else I am misunderstanding how the temp dir gets created and it's not a big deal.

**Comment by meteorchenwu** [comment]
It is great.I have merged it into my production environment.

**Comment by beliefer** [comment]
> OK it's probably that I don't know this code well. Maybe my question is this: when `isLocal == true`, `storage.locationUri` is a local path, so `targetPath` points to a local directory. It is now used as a path on distributed storage here. It may not exist or be usable that way. Is that OK? It may happen to work but seems a little wrong. Or else I am misunderstanding how the temp dir gets created and it's not a big deal.

Yes, when isLocal == true, storage.locationUri is a local path, so targetPath points to a local directory. It is not used as a path on distributed storage here.
You means targetPath may not exist or be usable that way, so this PR will create the local directory explicitly.
Let me explain in another way:
Driver and Executor running in different node(or machine).The SQL start with insert overwrite local directory will write data to a temp path first. The original code will use a local temp path.The local temp path is valid in node that running Driver and it is invalid in node that running Executor.
So this PR change the local temp path to a distributed storage path can be recognized by Driver and all the Executor.

**Comment by srowen** [comment]
Yes I get all that, but here:
`val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)`
`targetPath` is local. Why would you know it's valid path on distributed storage?

**Comment by beliefer** [comment]
> It is great.I have merged it into my production environment.

I add some code to guarantee the Robustness.

**Comment by beliefer** [comment]
> Yes I get all that, but here:
> `val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)`
> `targetPath` is local. Why would you know it's valid path on distributed storage?

In `yarn-client` mode,  `qualifiedPath` is a distributed path.
In `local` mode,  `qualifiedPath` is a local path.

**Comment by srowen** [comment]
After the change, `qualifiedPath` is constructed as:
`val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)`
Before it was qualified w.r.t. the local file system.
Doesn't this become a path on the distributed file system? that's the point of this change. But, if `targetPath` is "/tmp/foo", this assumes it's valid to use "/tmp/foo" on the distributed storage. That doesn't seem technically right. Am I missing something so far?

**Comment by beliefer** [comment]
> After the change, `qualifiedPath` is constructed as:
> `val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)`
> Before it was qualified w.r.t. the local file system.
> Doesn't this become a path on the distributed file system? that's the point of this change. But, if `targetPath` is "/tmp/foo", this assumes it's valid to use "/tmp/foo" on the distributed storage. That doesn't seem technically right. Am I missing something so far?

The key issue is not the `targetPath` but the `tmpPath`.
Before the change of this PR, the variables on our production environment in `local` mode as follows:
`targetPath`                                     `/home/***/gja_test/call_center`
`qualifiedPath`(not used in fact)     `hdfs://namenode.***:9000/home/***/gja_test/call_center`
`writeToPath`                                  `file:/home/***/gja_test/call_center`
`tmpPath`                                        `file:/home/***/hive/stagingdir_hive_2019-02-19_17-31-00_678_1816816774691551856-1/-ext-10000`

Before the change of this PR, the variables on our production environment in `yarn-client` mode as follows:
`targetPath`                                     `/home/***/gja_test/call_center`
`qualifiedPath`(not used in fact)     `hdfs://namenode.***:9000/home/***/gja_test/call_center`
`writeToPath`                                   `file:/home/***/gja_test/call_center`
`tmpPath`                                         `file:/home/***/hive/stagingdir_hive_2019-02-19_17-31-00_678_1816816774691551856-1/-ext-10000`

After the change of this PR , the variables on our production environment in `yarn-client` mode as follows:
`targetPath`                                 `/home/***/gja_test/call_center`
`qualifiedPath`                            `hdfs://namenode.***:9000/home/***/gja_test/call_center`
`writeToPath`                              `file:/home/***/gja_test/call_center`
`tmpPath`                                   `hdfs://namenode.***:9000/home/***/hive/stagingdir_hive_2019-02-19_17-18-29_777_5647666931466990865-1/-ext-10000`

**Comment by beliefer** [comment]
> After the change, `qualifiedPath` is constructed as:
> `val qualifiedPath = FileUtils.makeQualified(targetPath, hadoopConf)`
> Before it was qualified w.r.t. the local file system.
> Doesn't this become a path on the distributed file system? that's the point of this change. But, if `targetPath` is "/tmp/foo", this assumes it's valid to use "/tmp/foo" on the distributed storage. That doesn't seem technically right. Am I missing something so far?

If `targetPath` is "/tmp/foo", and assumes it's valid to use `tmpPath` "hdfs://tmp/foo" on the distributed storage. 
But the original code uses the "file:/tmp/foo" as `tmpPath`.

**Comment by srowen** [comment]
Yeah, I think that's a potential problem. We don't know whether that same path is valid on a distributed store. I agree there's a problem to fix here. Is there any standard tmp dir you can instead write to in this case rather than reusing that local path as a distributed path?

**Comment by beliefer** [comment]
> Yeah, I think that's a potential problem. We don't know whether that same path is valid on a distributed store. I agree there's a problem to fix here. Is there any standard tmp dir you can instead write to in this case rather than reusing that local path as a distributed path?

The purpose of executing this command is to write the data to a local path. So the `targetPath` should be a `local` path. The `writeToPath` created with `localFileSystem.makeQualified(targetPath)` used to confirm the final path which store result data. The `localFileSystem`'s type is `LocalFileSystem` so `writeToPath` is a local qualified `targetPath`. The `qualifiedPath` is always a distributed path in `yarn-client` deploy mode. The `tmpPath` is created with `getExternalTmpPath` and `qualifiedPath` and used as temporary storage of result data. I think `getExternalTmpPath` used to generate a string about standard tmp dir. If `qualifiedPath` is a local path, the `tmpPath` is created with `getExternalTmpPath` is a local path too. If `qualifiedPath` is a distributed path, the `tmpPath` is created with `getExternalTmpPath` is a distributed path too. The `qualifiedPath`  is not the same as `writeToPath` or `targetPath`，so there not reuse the local path.

**Comment by srowen** [comment]
OK, if you're saying this change doesn't cause the temp path to be some function of the (local) write path, then I get it. If it just comes up with a standard tmp location that is now on distributed storage, I get it.

**Comment by beliefer** [comment]
> OK, if you're saying this change doesn't cause the temp path to be some function of the (local) write path, then I get it. If it just comes up with a standard tmp location that is now on distributed storage, I get it.

Yes，it is.

**Comment by srowen** [comment]
@maropu @cloud-fan does this sound reasonable to you? given the extended discussion above and the last couple comments here

**Comment by beliefer** [comment]
@maropu @cloud-fan srowen and I have a lot discussion above,can you pay attention to last couple comments?

**Comment by beliefer** [comment]
@cloud-fan Please look this PR, thanks!

**Comment by beliefer** [comment]
> @maropu @cloud-fan does this sound reasonable to you? given the extended discussion above and the last couple comments here

This PR has been going on for more than 40 days. Please help me advance this PR.

**Comment by beliefer** [comment]
> @maropu @cloud-fan does this sound reasonable to you? given the extended discussion above and the last couple comments here

This PR has been going on for more than 40 days. Please help me advance this PR.

**Comment by srowen** [comment]
Given the discussion, my understanding, passed tests, and reports that this fixes the problem in a prod env, I'm merging this to master.

**Comment by beliefer** [comment]
> Given the discussion, my understanding, passed tests, and reports that this fixes the problem in a prod env, I'm merging this to master.

Thank you very mush.Your rigor and attitude towards the discussion is very helpful to the contributors.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala [hunk 9]
```diff
@@ -87,17 +87,17 @@ case class InsertIntoHiveDirCommand(
...
         val localFileSystem = FileSystem.getLocal(jobConf)
-        localFileSystem.makeQualified(targetPath)
+        (localFileSystem.makeQualified(targetPath), localFileSystem)
       } else {
-        val dfs = qualifiedPath.getFileSystem(jobConf)
-        if (!dfs.exists(qualifiedPath)) {
-          dfs.mkdirs(qualifiedPath.getParent)
-        }
-        qualifiedPath
+        val dfs = qualifiedPath.getFileSystem(hadoopConf)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: What is the reason we change this line from jobConf to hadoopConf?

Comment (gatorsmile) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: cc @beliefer @srowen

Comment (srowen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: Good question; I believe it was for consistency? the previous line used `hadoopConf`. `jobConf` inherits the same properties so I'd imagine it won't matter in this context. Does it cause an issue that you know of?

Comment (gatorsmile) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala]: It sounds like it is safe to remove the variable jobConf and replace all the usage in this file by hadoopConf

**Pull request closed without merging**